File: harvester/parsers/registry/handlers/preamble.py
"""Preamble element handlers for aanhef (preamble) sections.

These handlers process elements found in the aanhef (preamble) of Dutch laws,
including the royal introduction (wij), considerations (considerans),
and proclamation (afkondiging).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

from harvester.parsers.registry.protocols import (
    ElementType,
    ParseContext,
    ParseResult,
    RecurseFn,
    extract_text_with_tail,
)

if TYPE_CHECKING:
    from lxml import etree


@dataclass
class WijHandler:
    """Handler for <wij> (royal introduction) elements.

    Contains the formal royal introduction text like
    "Wij Beatrix, bij de gratie Gods, Koningin der Nederlanden..."
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.INLINE

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        text = elem.text.strip() if elem.text else ""
        return ParseResult(text=text)


@dataclass
class ConsideransHandler:
    """Handler for <considerans> (considerations) elements.

    Contains the considerations/recitals of the law. Processes all
    <considerans.al> child elements.
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.STRUCTURAL

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        parts: list[str] = []

        for al in elem.findall(".//considerans.al"):
            result = recurse(al, context)
            if result.text:
                parts.append(result.text)

        return ParseResult(text="\n\n".join(parts))


@dataclass
class ConsideransAlHandler:
    """Handler for <considerans.al> (consideration paragraph) elements.

    Individual paragraph within the considerations section.
    Processes inline content including references.
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.INLINE

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        return ParseResult(text=extract_text_with_tail(elem, context, recurse))


@dataclass
class AfkondigingHandler:
    """Handler for <afkondiging> (proclamation) elements.

    Contains the formal proclamation text. Processes all <al> child elements.
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.STRUCTURAL

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        parts: list[str] = []

        for al in elem.findall(".//al"):
            result = recurse(al, context)
            if result.text:
                parts.append(result.text)

        return ParseResult(text="\n\n".join(parts))


@dataclass
class AanhefHandler:
    """Handler for <aanhef> (preamble) elements.

    The preamble contains the royal introduction, considerations,
    and proclamation. Processes all child elements in order.
    """

    @property
    def element_type(self) -> ElementType:
        return ElementType.STRUCTURAL

    def can_handle(self, elem: etree._Element, context: ParseContext) -> bool:
        return True

    def handle(
        self,
        elem: etree._Element,
        context: ParseContext,
        recurse: RecurseFn,
    ) -> ParseResult:
        parts: list[str] = []

        # Process wij
        wij_elem = elem.find("wij")
        if wij_elem is not None:
            result = recurse(wij_elem, context)
            if result.text:
                parts.append(result.text)

        # Process considerans
        considerans_elem = elem.find("considerans")
        if considerans_elem is not None:
            result = recurse(considerans_elem, context)
            if result.text:
                parts.append(result.text)

        # Process afkondiging
        afkondiging_elem = elem.find("afkondiging")
        if afkondiging_elem is not None:
            result = recurse(afkondiging_elem, context)
            if result.text:
                parts.append(result.text)

        return ParseResult(text="\n\n".join(parts))

File: packages/harvester/src/registry/handlers/preamble.rs
//! Preamble element handlers for law introductions.
//!
//! These handlers process elements in the preamble (aanhef) of Dutch laws,
//! including the royal introduction and considerations.

use roxmltree::Node;

use crate::registry::handler::{extract_text_with_tail, ElementHandler, RecurseFn};
use crate::registry::types::{ElementType, ParseContext, ParseResult};
use crate::xml::get_tag_name;

/// Handler for `<wij>` (royal "We") elements.
///
/// The opening of the preamble, typically "Wij Beatrix..." or "Wij Willem-Alexander...".
pub struct WijHandler;

impl ElementHandler for WijHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Inline
    }

    fn handle<'a, 'input>(
        &self,
        node: Node<'a, 'input>,
        _context: &mut ParseContext<'_>,
        _recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        let text = node.text().map(|s| s.trim()).unwrap_or_default();
        ParseResult::new(text)
    }
}

/// Handler for `<considerans>` (considerations) elements.
///
/// Contains the legal considerations for the law.
pub struct ConsideransHandler;

impl ElementHandler for ConsideransHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Structural
    }

    fn handle<'a, 'input>(
        &self,
        node: Node<'a, 'input>,
        context: &mut ParseContext<'_>,
        recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        let mut parts: Vec<String> = Vec::new();

        for child in node.children() {
            if child.is_element() {
                let result = recurse(child, context);
                if !result.text.is_empty() {
                    parts.push(result.text);
                }
            }
        }

        ParseResult::new(parts.join("\n\n"))
    }
}

/// Handler for `<considerans.al>` (consideration paragraph) elements.
pub struct ConsideransAlHandler;

impl ElementHandler for ConsideransAlHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Inline
    }

    fn handle<'a, 'input>(
        &self,
        node: Node<'a, 'input>,
        context: &mut ParseContext<'_>,
        recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        ParseResult::new(extract_text_with_tail(node, context, recurse))
    }
}

/// Handler for `<afkondiging>` (proclamation) elements.
pub struct AfkondigingHandler;

impl ElementHandler for AfkondigingHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Structural
    }

    fn handle<'a, 'input>(
        &self,
        node: Node<'a, 'input>,
        context: &mut ParseContext<'_>,
        recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        let mut parts: Vec<String> = Vec::new();

        for child in node.children() {
            if child.is_element() && get_tag_name(child) == "al" {
                let result = recurse(child, context);
                if !result.text.is_empty() {
                    parts.push(result.text);
                }
            }
        }

        ParseResult::new(parts.join("\n\n"))
    }
}

/// Handler for `<aanhef>` (preamble) elements.
pub struct AanhefHandler;

impl ElementHandler for AanhefHandler {
    fn element_type(&self) -> ElementType {
        ElementType::Structural
    }

    fn handle<'a, 'input>(
        &self,
        node: Node<'a, 'input>,
        context: &mut ParseContext<'_>,
        recurse: &RecurseFn<'a, 'input>,
    ) -> ParseResult {
        let mut parts: Vec<String> = Vec::new();

        for child in node.children() {
            if child.is_element() {
                let result = recurse(child, context);
                if !result.text.is_empty() {
                    parts.push(result.text);
                }
            }
        }

        ParseResult::new(parts.join("\n\n"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use roxmltree::Document;

    fn parse_and_handle<H: ElementHandler>(handler: &H, xml: &str) -> ParseResult {
        let doc = Document::parse(xml).unwrap();
        let node = doc.root_element();
        let mut context = ParseContext::new("BWBR0000000", "2025-01-01");
        let recurse = |_: Node<'_, '_>, _: &mut ParseContext<'_>| ParseResult::empty();
        handler.handle(node, &mut context, &recurse)
    }

    #[test]
    fn test_wij_handler() {
        let result = parse_and_handle(
            &WijHandler,
            "<wij>Wij Willem-Alexander, bij de gratie Gods, Koning der Nederlanden</wij>",
        );
        assert!(result.text.contains("Willem-Alexander"));
    }

    #[test]
    fn test_considerans_al_handler() {
        let result = parse_and_handle(
            &ConsideransAlHandler,
            "<considerans.al>Overwegende dat...</considerans.al>",
        );
        assert!(result.text.contains("Overwegende"));
    }
}
